### PR TITLE
Enable setting client.path.prefix='/'

### PR DESCRIPTION
**Pull request by beiske**
These are changes for running the rest tests against a cluster on Cloud.

* Ensure all paths have a leading `/` by setting path prefix
* ~Allow adding additional blacklisted tests for `CoreWithSecurityClientYamlTestSuiteIT`~ (Merged in #31074)
* ~Allow specifying custom cluster credentials for `CoreWithSecurityClientYamlTestSuiteIT`~ (Merged in #31235)

### CODE REVIEW

**Comment by javanna** [comment]
As far as I can see this change makes it possible to configure '/' as a `pathPrefix` in the low-level REST client. I wonder if this is how we want to make this change. We never modified paths to start with '/' but Elasticsearch used to be lenient there and accept relative paths as they were absolute. Considering that Elasticsearch no longer makes relative paths work as if they were absolute, should we rather always add the '/' when missing on any path, rather than adding the ability to configure '/' as `pathPrefix`?

**Comment by beiske** [comment]
>...should we rather always add the '/' when missing on any path, rather than adding the ability to configure '/' as pathPrefix?

Sounds good to me. That was one of my original proposals. Using the prefix path was done by proposal from @jasontedor. Maybe he was anticipating issues when backporting to 6.x or something else I don't remember?

**Comment by jasontedor** [comment]
I really do not like manipulating user input. If the user says: send the request to a/b/c we should send the request to a/b/c.

**Comment by javanna** [comment]
@jasontedor I agree on the sentiment. On my end, I would like to figure out what Elasticsearch accepts with the recent changes in master, what works and what doesn't, and make sure that we are heading in the right direction. I would find it a bit weird to allow users to provide relative paths if they don't work at all, so wondering if this is only a proxy problem or not. An alternative could be to consider to no longer accept relative paths in the client, assuming that recent changes in the Elasticsearch server code doesn't make them work well like it did before.

File: client/rest/src/main/java/org/elasticsearch/client/RestClient.java [hunk 1]
```diff
@@ -581,16 +581,14 @@ static URI buildUri(String pathPrefix, String path, Map<String, String> params)
...
-                } else {
-                    fullPath = pathPrefix + "/" + path;
-                }
+            if (pathPrefix != null && !pathPrefix.isEmpty()) {
+                fullPath = (pathPrefix + "/" + path).replaceAll("//+", "/");
             } else {
                 fullPath = path;
             }
 
+            logger.debug("Building URI with fullpath: ["+ fullPath+"] for path: ["+path+"]");
```
Comment (javanna) [inline client/rest/src/main/java/org/elasticsearch/client/RestClient.java]: nit: I wonder if this new log line is necessary, given that it's called for every single request I'd try to do without it. If users want to log this though they can enable request tracing?

**Comment by javanna** [comment]
I had another look and I am ok with this approach. I can't comment on the build changes though.

**Comment by beiske** [comment]
Jenkins test this

**Comment by suyograo** [comment]
> I can't comment on the build changes though.

@jasontedor could you please review or assign this to someone who can review the build changes bits?

**Comment by jasontedor** [comment]
This will be looked at soon; we are bottlenecked right now.

**Comment by jasontedor** [comment]
I want to break this PR into three separate PRs as there are multiple changes bundled here.

**Comment by jasontedor** [comment]
The first PR is #30885.

**Comment by jasontedor** [comment]
A second PR is #31074.

**Comment by jasontedor** [comment]
A third PR is #31235.

**Comment by beiske** [comment]
I have rebased this one after the merges of #30885, #31074 and #31235 so that it now only contains the path prefix changes.

**Review by jasontedor — CHANGES_REQUESTED** [review]
Looks good in general but I left some comments.

File: client/rest/src/main/java/org/elasticsearch/client/RestClient.java [hunk 2]
```diff
@@ -796,12 +796,8 @@ static URI buildUri(String pathPrefix, String path, Map<String, String> params)
...
         try {
             String fullPath;
-            if (pathPrefix != null) {
-                if (path.startsWith("/")) {
-                    fullPath = pathPrefix + path;
-                } else {
-                    fullPath = pathPrefix + "/" + path;
-                }
+            if (pathPrefix != null && !pathPrefix.isEmpty()) {
+                fullPath = (pathPrefix + "/" + path).replaceAll("//+", "/");
```
[the remark this answers is not in the document]
Comment (beiske) [reply client/rest/src/main/java/org/elasticsearch/client/RestClient.java]: A `pathPrefix` that is a single slash, would also be a `pathPrefix` that ends with a slash. So a naturalæ consequence of using the `pathPrefix` to ensure there always is a leading slash is allowing `pathPrefix` with a trailing slash. Of course one could make a special case for trailing slash when it is the entire prefix, but that seemed like very circumstantial logic to me.

File: client/rest/src/main/java/org/elasticsearch/client/RestClient.java [hunk 3]
```diff
@@ -796,8 +796,11 @@ static URI buildUri(String pathPrefix, String path, Map<String, String> params)
         Objects.requireNonNull(path, "path must not be null");
         try {
             String fullPath;
-            if (pathPrefix != null) {
-                if (path.startsWith("/")) {
+            if (pathPrefix != null && pathPrefix.isEmpty() == false) {
+                fullPath = (pathPrefix + "/" + path).replaceAll("//+", "/");
```
Comment (jasontedor) [inline client/rest/src/main/java/org/elasticsearch/client/RestClient.java]: This value is immediately written over in any one of the following `if`, `else if`, or `else` blocks so this line does not do anything?

**Review by jasontedor — CHANGES_REQUESTED** [review]
I left a comment.

**Review by jasontedor — APPROVED** [review]

**Comment by beiske** [comment]
@jasontedor Thank you!

**Comment by aug24** [comment]
I've done some more digging, and I would like to discuss this further if that's OK?  I believe the current position is inconsistent.

In RestClientBuilder.setPathPrefix, there is the following code:

        if (cleanPathPrefix.startsWith("/") == false) {
            cleanPathPrefix = "/" + cleanPathPrefix;
        }

ie paths are forced to have a leading slash.

There is also this:

        if (cleanPathPrefix.isEmpty() || "/".equals(cleanPathPrefix)) {
            throw new IllegalArgumentException("pathPrefix must not be empty or '/': [" + pathPrefix + "]");
        }

ie you cannot simply handle the trivial case of "always prepend a slash" because the code actually doesn't let you.

As far as I can tell, there is no correct use case for the user providing a relative path URI without a pathPrefix.  If we don't automatically fix this error (per the fix I suggested) then we should throw an exception with advice that either a pathPrefix or a leading slash is required.

**Comment by aug24** [comment]
Apologies: more context is at https://github.com/elastic/elasticsearch/pull/32714.  Perhaps this discussion should continue there.  I'll cross-post.

I clearly have a slightly old client as the second block, which was removed on July 1st, is present.  

In short: the rest client builder permits the user to build a request which is, as I understand it, invalid.  

I suggest adding a leading slash if one was not provided, or throwing an exception, as I think a request without a leading is slash is invalid (and so do AWS ELBs, which leads to a hard to debug error).

I notice that we do force a leading slash if a pathPrefix without one is given.  In fact the comment on this PR:

> Some proxies require all requests to have paths starting with / since
> there are no relative paths at the HTTP connection level. Elasticsearch
> assumes paths are absolute. 

is precisely the reason why I think the other PR should be accepted.

**Pull request merged**